app/src/main/java/com/blogspot/e_kanivets/moneytracker/adapter/RecordReportAdapter.kt

```diff
@@ -0,0 +1,150 @@
+        val posWithoutSummary = getPosWithoutSummary(position)
+        if (posWithoutSummary < 0) return
+
+        if (holder is ChildViewHolder) {
+            val row = items[posWithoutSummary] as RecordReportItem.ChildRow
+            holder.tvCategory.text = row.title
+            holder.tvTotal.text = row.amount
+            holder.tvTotal.setTextColor(if (row.amount.first() != '-') green else red)
+        } else {
+            val parentViewHolder = holder as ParentViewHolder
+            val row = items[posWithoutSummary] as RecordReportItem.ParentRow
+            parentViewHolder.tvCategory.text = row.category
+            parentViewHolder.tvTotal.text = row.totalAmount
+            holder.tvTotal.setTextColor(if (row.totalAmount.first() != '-') green else red)
+        }
+    }
+
+    fun setData(items: MutableList<RecordReportItem>, data: HashMap<RecordReportItem.ParentRow, List<RecordReportItem.ChildRow>>) {
+        this.items = items
+        this.data = data
+        notifyDataSetChanged()
+    }
+
+    fun addSummaryView(summaryView: View) {
+        this.summaryViewHolder = summaryView.tag as RecyclerView.ViewHolder
+    }
+
+    private fun changeItems(position: Int) {
+        if (items[position] is RecordReportItem.ParentRow) {
+            val parentRow: RecordReportItem.ParentRow = items[position] as RecordReportItem.ParentRow
+
+            if (parentRow.isOpen) {
+                val item = items.filterIndexed { index, _ -> index > position }
+                        .find { recordReportItem -> recordReportItem is RecordReportItem.ParentRow }
+
+                val lastChildInd = if (item != null) items.indexOf(item) else items.size
+
+                items.subList(position + 1, lastChildInd).clear()
+                val itemCount = lastChildInd - position - 1
+                notifyItemRangeRemoved(getPosWithSummary(position + 1), itemCount)
+
+                parentRow.isOpen = false
+            } else {
+                data[parentRow]?.let { childRows ->
+                    var lastChildInd = position + 1
+                    for (childRow in childRows) {
+                        items.add(lastChildInd, childRow)
+                        lastChildInd++
+                    }
+                    notifyItemRangeInserted(getPosWithSummary(position + 1), childRows.size)
+                }
+
+                parentRow.isOpen = true
+            }
+        }
+    }
+
+    private fun getPosWithSummary(position: Int): Int {
+        return position + 1
+    }
+
+    private fun getPosWithoutSummary(position: Int): Int {
+        return position - 1
+    }
+
+    inner class ParentViewHolder : RecyclerView.ViewHolder {
+
+        var tvCategory: TextView
+        var tvTotal: TextView
+
+        constructor(view: View) : super(view) {
+            tvCategory = view.tvCategory
+            tvTotal = view.tvTotal
+
+            view.setOnClickListener {
+                if (view.lowerDivider.visibility == View.VISIBLE) {
```

Comment:
Relying on view visibility isn't the best idea. It's better to introduce a class field for that.